### ISSUE TITLE
Alternative definition of binomial coefficients for negative arguments and associated documentation

### ISSUE THREAD

**Issue by jeremies** [sympy/sympy#14577]
According to this paper (https://arxiv.org/pdf/1105.3689.pdf), if we define binomial coefficients `binomial(n,k)` in terms of gamma functions, then it follows formula (2.1) of the paper for negative integer n and integer k, that for example gives binomial(-1,-1) as 1 in contradiction with SymPy that gives 0. This is the definition taken by the Wolfram Language (http://mathworld.wolfram.com/BinomialCoefficient.html). Here is how the computation is done (https://www.johndcook.com/blog/computing_binomial_coefficients/). Should we follow this definition, since we are defining binomial with gamma functions?

SymPy currently defines for the sake of convenience for negative integer 'k' that this function will return zero no matter what valued is the other argument. Why? Can this be more precisely defined as the definition in http://functions.wolfram.com/GammaBetaErf/Binomial/02/ ?

**Comment by sidhantnagpal** [comment]
> Should we follow this definition, since we are defining binomial with gamma functions?

Actually, the result `0` obtained using the current definition ([Definition 3](https://www.johndcook.com/blog/binomial_coefficients/)) of master also uses `gamma functions` with `repeated limits` (the order of limits define the result).

I am not sure if it is ideal to cancel singularities as mentioned in [arXiv:1105.3689](https://arxiv.org/pdf/1105.3689.pdf). What are the advantages of this method?

Interestingly, this wolfram site (http://functions.wolfram.com/GammaBetaErf/Binomial/02/) that you mentioned, defines the value of `C(n, k)` to be `0` for integers `k <= n < 0`.

**Comment by jksuom** [comment]
I think that the current docstring is quite informative:

    """Implementation of the binomial coefficient. It can be defined
    in two ways depending on its desired interpretation:

        C(n,k) = n!/(k!(n-k)!)   or   C(n, k) = ff(n, k)/k!

    First, in a strict combinatorial sense it defines the
    number of ways we can choose 'k' elements from a set of
    'n' elements. In this case both arguments are nonnegative
    integers and binomial is computed using an efficient
    algorithm based on prime factorization.

    The other definition is generalization for arbitrary 'n',
    however 'k' must also be nonnegative. This case is very
    useful when evaluating summations.

    For the sake of convenience for negative integer 'k' this function
    will return zero no matter what valued is the other argument.

It starts with the original definition with two positive integers `n` and `k`, and then goes on to introduce the generalization where `n` is allowed to be any complex number. Only some justification would be needed for the last paragraph. ` C(n, k) = ff(n, k)/k!` is well defined for positive integers. If we want to define it for negative integers `k`, we can first extend it fairly naturally to a meromorphic function of two variables ('rewrite(gamma)'). I think that the most natural extension to negative integers `k` is then obtained by simply restricting this function. For each negative integer `k`, it is a meromorphic function of a single variable `n`. This function is identically zero. (It has removable singularities at isolated points.)

**Comment by jeremies** [comment]
> What are the advantages of this method?

I'm not very familiar with binomials for negative integer `n` and integer `k`, but the paper in https://arxiv.org/pdf/1105.3689.pdf says that the formula (2.1) is in agreement with the binomial theorem and with combinatorial identities with a few special exceptions. Moreover, definition (2.1) define the binomial coefficient as a continuous function for all complex (including all integer) arguments, except for negative integer x and non-integer y, in which case the binomial coefficient is infinite. In particular it will be continuous at negative integer `n` and integer `k`, for example now SymPy gives `binomial(-5,-10)` as `0` and `binomial(-5.0001,-10.0001)` as `-126.009395277087` but with the definition of the paper it will give `binomial(-5,-10)` as `-126` and `binomial(-5.0001,-10.0001)` as `-126.009395277087`.

> I think that the most natural extension to negative integers k is then obtained by simply restricting this function.

What other mathematical software uses this definition? What are the advantages of this definition?

**Comment by jeremies** [comment]
@asmeurer @jksuom @sidhantnagpal 
In particular if we accept `binomial(-p,-p)` for `p` integer and positive to be zero then `combsimp(binomial(1,p))` gives `zoo/(p-1)` in pull request #14557 that is wrong. I'm trying to repair the issue #14528 but with the recent pull request #14575 merged this is impossible since now `binomial(-p,-p)` for `p` integer and positive gives zero. Before this pull request the result of `binomial(-p,-p)` was 1 and then my pull request #14557 give the correct result `1/(p!(1-p)!)`. I think that the definition of binomial of the paper https://arxiv.org/pdf/1105.3689.pdf should be implemented. At least it would give correct simplifications.

**Comment by jeremies** [comment]
@asmeurer @jksuom @sidhantnagpal
Sorry if the later message wasn't not so much clear. This is the problem, if you execute the following in my branch named issue_14528_binomial it gives the following (before pull request #14575, this was correct):
``` python
>>> p = symbols("p", positive=True, integer=True)
>>> combsimp(binomial(1, p))
zoo/(p - 1)
```

**Comment by asmeurer** [comment]
> In particular it will be continuous at negative integer n and integer k, for example now SymPy gives binomial(-5,-10) as 0 and binomial(-5.0001,-10.0001) as -126.009395277087 but with the definition of the paper it will give binomial(-5,-10) as -126 and binomial(-5.0001,-10.0001) as -126.009395277087.

This is misleading. The current definition is also based on a limit, but for k first (see the John Cook post). For instance:

```
>>> binomial(-5.0001,-10.0000000001)
-0.000126002532276851
```

If the negative integers were removable singularities, there would be no issue here. But they depend on which variable is approached first. If I am understanding it correctly, it's a matter of approaching n first or k first. 

Here's my interpretation of the problem. If you take `binomial(n, k)` to be the coefficient of `x**k` in `(1 + x)**n` (which is the base definition [Wikipedia](https://en.wikipedia.org/wiki/Binomial_coefficient) uses), then as a natural extension for negative `n`, we have

```
>>> (1/(x + 1)).series()
1 - x + x**2 - x**3 + x**4 - x**5 + O(x**6)
```

Meaning the coefficient of `1/x` in `1/(x + 1)` is `0`. 

However, as pointed out by the paper, if we instead look at `(x + y)**n` and consider the coefficient of `x**k*y**(n - k)`there are two series, depending if we expand with respect to `x` or `y`:

```
>>> (1/(x + y)).series(x)
1/y - x/y**2 + x**2/y**3 - x**3/y**4 + x**4/y**5 - x**5/y**6 + O(x**6)
>>> (1/(x + y)).series(y)
-y**5/x**6 + y**4/x**5 - y**3/x**4 + y**2/x**3 - y/x**2 + 1/x + O(y**6)
```

In the first one, the coefficient of `x**-1*y**(-1 - -1)`, that is, `1/x` is `0`. In the second one, it is `1`. The difference, as pointed out in the paper, is the values of `x` and `y` for which the series converge.

Thus, we can chose this definition and still be consistent with the binomial theorem in some sense.

Another way to look at it is that the series for `1/(x + 1)` only converges for `|x| < 1`, so to consider `|x| > 1`, we should instead look at `1/(1/x + 1)`

```
>>> (1/(1/x + 1)).series()
x - x**2 + x**3 - x**4 + x**5 + O(x**6)
```

Here the coefficient of `x` (corresponding to `binomial(-1, -1)`, since we inverted `x` in the expansion) is `1`.

The paper doesn't, as far as I can tell, consider the alternative definition of letting `binomial(n, k) == 0` for all negative integer `k`, so there aren't really arguments there why one definition is better than the other. 

The above series expansion argument does seem a bit fishy to me. To expand the definition for negative powers, we have to consider a different radius of convergence depending on whether we expand around x=0 or x=infinity, namely, picking the one where the coefficient in question is not 0. 

Although this does seem to be how things work for nonintegral values:

```
>>> (1/sqrt(1/x + 1)).series()
sqrt(x) - x**(3/2)/2 + 3*x**(5/2)/8 - 5*x**(7/2)/16 + 35*x**(9/2)/128 - 63*x**(11/2)/256 + O(x**6)
>>> (1/sqrt(x + 1)).series()
1 - x/2 + 3*x**2/8 - 5*x**3/16 + 35*x**4/128 - 63*x**5/256 + O(x**6)
>>> binomial(-S(1)/2, -1)
0
>>> binomial(-S(1)/2, -S(1)/2)
1
>>> binomial(-S(1)/2, 1)
-1/2
```

Is there a more reasoned motivation for doing things one way or the other, at least in terms of generalizing the definition in terms of series expansions?

**Comment by sidhantnagpal** [comment]
Yes, the current behaviour is consistent with Pascal's Identity and which uses the definition of Repeated Limit and Gamma Function (Symmetry Identity does not hold by the virtue of the application of limit).

The insight for realizing binomial coefficients using series expansions is interesting, although as mentioned the constraints on convergence again cause an ambiguity in which definition is useful. It is intriguing that both results can be realized using this argument.

**Comment by jeremies** [comment]
I like the motivations of the current definition given by asmeurer. But here are some problems that I found with the current definition regarding simplifications.

The simplifications actually done with SymPy, that are in written in the issue #14612 are wrong, and could be repaired with the definition of the paper. 

More important is the fact that function _gammasimp used one of the previous identities, `rf(n,k) = gamma(k+1) * binomial(n+k-1,k)`, and with the current definition of binomial I was forced in pull request #14557 to replace that identity with the following identity `rf(n,k) = gamma(k+1)` that is true in the context that it is used (when `n = 1`) but not always. This identity in _gammasimp is important because if one deletes it it doesn't simplify correctly some expressions.

To view the decision taken by other software regarding binomials I found that there is software like Maple and Mathematica that define binomial according to the paper and there is software like Sage and Maxima that implement the current definition binomial of SymPy. There is no perfect definition each one has its advantages.

**Comment by asmeurer** [comment]
This was changed since the last release, so we should probably decide what we want to do before we release again. I'm leaning toward our current definition. 

In sympy-1.1.1, `binomial(-1, -1)` gives `1`. Was the definition there the one from the paper, or was it just wrong?

**Comment by sidhantnagpal** [comment]
I am +1 for this:
> Here's my interpretation of the problem. If you take `binomial(n, k)` to be the coefficient of `x**k` in `(1 + x)**n` (which is the base definition Wikipedia uses), then as a natural extension for negative `n`, we have
> ```
> >>> (1/(x + 1)).series()
> 1 - x + x**2 - x**3 + x**4 - x**5 + O(x**6)
> ```
> Meaning the coefficient of `1/x` in `1/(x + 1)` is `0`.

**Comment by sidhantnagpal** [comment]
As mentioned in John Cook's post, the definition for complex values extends using Gamma Function. [Wikipedia](https://en.wikipedia.org/wiki/Binomial_coefficient#Two_real_or_complex_valued_arguments) seems to include an explanation for that as well.

**Comment by asmeurer** [comment]
My question is, is the extension unique? I've never really clear on how uniqueness for analytic continuations works for extending a function defined on Z to a function defined on C.

**Comment by jksuom** [comment]
An analytic extension of a function defined on Z need not be unique. The extension can always be modified by adding a functions that is zero on Z such as sin πx. For uniqueness of the extension, it is sufficient that the set on which the function is defined has at least one accumulation point (in each component of the domain of the extension if that is not connected).

**Comment by asmeurer** [comment]
That's something that has never been clear to me. How is, for instance, `gamma(z + 1)` the "proper" analytic continuation of `n!` and not `gamma(z + 1) + sin(z*pi)` or `cos(2*z*pi)*gamma(z + 1)`? Is there some way to use the "accumulation point" at infinity? Or is it just strictly convention?

**Comment by jksuom** [comment]
It is a convention, but there are good grounds for it. For example, `gamma(z + 1)` is the only analytic function extending the factorial that is *logarithmically convex* on the positive real line, i.e., the second derivative of its logarithm is positive. That was first pointed out by E. Artin in his book *Einführung in die Theorie der Gammafunktion* (1931).

**Comment by asmeurer** [comment]
Interesting. So this also implies that the gamma formula for binomial is log convex wrt k (see also https://www.johndcook.com/blog/2013/05/21/binomial-coefficients-log-concave/). Assumedly gamma(z + 1)/(gamma(z - k + 1)*gamma(k + 1)) is also unique in this sense for any given (complex) z.

**Comment by iammosespaulr** [comment]
> A recent paper about dealing with and interpreting negative arguments is [here](https://arxiv.org/pdf/1802.02684.pdf). I think it would be worth allowing binomial to return values for all arbitrary arguments. (And perhaps doing so before committing #18960.)
What about this ?

**Comment by asmeurer** [comment]
I haven't looked at that yet. Can you summarize what it says?

**Comment by iammosespaulr** [comment]
![Screen Shot 2020-04-01 at 1 50 49 AM](https://user-images.githubusercontent.com/28682735/78071571-65aaea80-73bb-11ea-8ba5-85da766a3730.png)
eq(5) is what @jeremies had meant earlier. 😄 

> > What are the advantages of this method?
>In particular it will be continuous at negative integer `n` and integer `k`, for example now SymPy gives `binomial(-5,-10)` as `0` and `binomial(-5.0001,-10.0001)` as `-126.009395277087` but with the definition of the paper it will give `binomial(-5,-10)` as `-126` and `binomial(-5.0001,-10.0001)` as `-126.009395277087`.

**Comment by iammosespaulr** [comment]
>If the negative integers were removable singularities, there would be no issue here. But they depend on which variable is approached first. If I am understanding it correctly, it's a matter of approaching n first or k first.

Now they are removable singularities. eta in eq(5) enables that. Approaching via n or k wouldn't be necessary right?

>This is misleading. The current definition is also based on a limit, but for k first (see the John Cook post). For instance:

I believe John Cooks's "Definition 3: most general" is not the only way to approach the binomial coefficients of -ve Integers as a limit. 
![Screen Shot 2020-04-01 at 2 06 49 AM](https://user-images.githubusercontent.com/28682735/78072915-85dba900-73bd-11ea-920f-ef55499c2e31.png)

**Comment by iammosespaulr** [comment]
[and it also seems that mathematica uses the earlier definition mentioned in paper 1](https://reference.wolfram.com/language/ref/Binomial.html)
![Screen Shot 2020-04-01 at 2 08 50 AM](https://user-images.githubusercontent.com/28682735/78073163-faaee300-73bd-11ea-8444-ac8cb71d6914.png)

**Comment by iammosespaulr** [comment]
> By the way, the fact that the symmetry identity `binomial(n, k) = binomial(n, n - k)` remains valid for negative `n` and `k` seems to be a plus. That does not hold for the current code, for instance, for `binomial(-1, -1)`.
> 
> On the other hand, the addition identity `binomial(n, k) = binomial(n - 1, k) + binomial(n - 1, k - 1)` doesn't hold with the paper definition for `n = k = 0` (it should hold in every case for the "negative k gives 0" definition, no?).
> 
> The addition identity feels more important to me, as this is a defining identity (this is how Pascal's triangle is constructed). But what I'd like to get a feel for is how important these identities, and the exceptions therein, are to simplification in SymPy.  
----
![Screen Shot 2020-04-01 at 2 16 56 AM](https://user-images.githubusercontent.com/28682735/78073654-eddebf00-73be-11ea-99eb-d39b9d9a3b71.png)  

...

![Screen Shot 2020-04-01 at 2 20 51 AM](https://user-images.githubusercontent.com/28682735/78073968-72c9d880-73bf-11ea-99ac-3abfb97bfbe3.png)
They hold now!

**Comment by smichr** [comment]
btw, `sgn(k)` in SymPy can be written `Heaviside(k, H0=1)*binomial(k - n - 1, k)`

**Comment by iammosespaulr** [comment]
I've been working on a suitable implementation for that and there seems to be a problem with the limiting function for gamma
```python3
>>> x = symbols('x')
>>> limit(gamma(x)/(gamma(x - 1)*gamma(x + 2)), x, 0)
Limit(gamma(x)/(gamma(x - 1)*gamma(x + 2)), x, 0)
>>> limit(gamma(x - 1)/(gamma(x - 2)*gamma(x + 2)), x, 0)
Limit(gamma(x - 1)/(gamma(x - 2)*gamma(x + 2)), x, 0)
>>> 
```
but they are rather well established 
[limit(gamma(x)/(gamma(x - 1)*gamma(x + 2)), x, 0)](https://www.wolframalpha.com/input/?i=Limit%28gamma%28x%29%2F%28gamma%28x+-+1%29*gamma%28x+%2B+2%29%29%2C+x%2C+0%29)  
[limit(gamma(x - 1)/(gamma(x - 2)*gamma(x + 2)), x, 0)](https://www.wolframalpha.com/input/?i=limit%28gamma%28x+-+1%29%2F%28gamma%28x+-+2%29*gamma%28x+%2B+2%29%29%2C+x%2C+0%29)

**Comment by pitman** [comment]
I'm not a SymPy contributor, or even user (yet) but a mathematician who is familiar both with the combinatorics of binomial coefficients and their implementation in Mathematica, which I regard as the correct one. Independently of this conversation, I started writing something to clarify the confusion in the literature about the definition of binomial(n,k) for integers n and k.  The classical text of Jordan on Calculus of Variations (page 68) defines binomial(n,k) for all integers (n,k) in the  way described by iammosespaulr   above and implemented in Mathematica. This definition also appears on page 5 of Riordan's Introduction to Combinatorial Analysis, and was rediscovered by various later authors including Parker (1973), Loeb, and Sprugnoli as cited by the excellent recent paper on  Gaussian binomial coefficients with negative arguments, quoted by  iammosespaulr  and by the authority on binomial identities H. W. Gould in his review of Parker's 1973  paper in Mathematical Reviews MR0340034.
It is easily shown that this definition is uniquely determined by binomial(n,0) = binomial(n,n) = 1 fpr all integers n and the Pascal recursion at all (n,k) except (0,0) where 
1 + 1  =  binomial(-1,-1) + binomial(-1,0)  does not equal binomial (0,0) = 1.
This definition has many good properties presented by Loeb, including the reflection symmetry binomial(n,k)  = binomial(n, n-k) for all integers (n,k), which more than compensates for the
exception to the Pascal recursion at (0,0). If instead, you insist as Feller, Comtet and Knuth have done in their texts, that binomial(n,k) = 0 for k < 0, then you get Pascal recursion at all pairs of integers (n,.k), but the reflection formula fails.
For these reasons I strongly advocate for the Jordan/Riordan/Loeb definition of binomial(n,k)
for integers n and k,  and against the Feller/Comtet/Knuth definition, which if you accept the JRL definition just becomes binomial(n,k) (k > - 1) in Iverson notation.

**Comment by asmeurer** [comment]
@pitman what is the reasoning behind the reflection formula being more important than the Pascal one? For SymPy, if we have a symbolic expression with terms like binomial(n, k) without any assumptions on n and k, and we want to simplify it by making use of binomial identities, we cannot apply an identity unless it is true for all possible n and k. So for me, the question is, which identities are most important for simplification. 

Looking through the code, it's not obvious to me how this applies for `combsimp`, because it first rewrites to gammas and then does simplification based on gamma identities. Has anyone dug into this yet?

**Comment by iammosespaulr** [comment]
> The above series expansion argument does seem a bit fishy to me. To expand the definition for negative powers, we have to consider a different radius of convergence depending on whether we expand around x=0 or x=infinity, namely, picking the one where the coefficient in question is not 0.

Does this help ?
these are the only extreme limits that can be applied.
![Screen Shot 2020-04-23 at 3 10 33 AM](https://user-images.githubusercontent.com/28682735/80036850-2bdb8880-8510-11ea-9077-58b38b7ca40a.png)

**Comment by pitman** [comment]
I think the right thing to do is to warn or inform user that there is an issue about defn of
binom{n}{k} about which authorities differ according to whether binom{-1}{-1} = 1 or 0.
Do you want reflection to trump addition. or addition to trump reflection. You can't have both for all integer arguments.
Mathematica has opted for the 1 choice, and I think thats a good one. Mainly because if you want the 0 choice you can easily change the function by multiplying by a 1/0 function. There is no such simple thing the other way.

**Comment by iammosespaulr** [comment]
@asmeurer I took a deeper look at the 'Series Expansion` interpretation of binomial
![Screen Shot 2020-04-23 at 12 55 15 PM](https://user-images.githubusercontent.com/28682735/80073859-07aa9680-8566-11ea-8bd8-e70317745e28.png)

If binomials is extended for -ve ints then,
```python3
coeff(x)              coeff(y)
binomial( -2, 1) == binomial( -2, -3) ( -2)
binomial( -2, 2) == binomial( -2, -4) ( 3)
binomial( -2, 3) == binomial( -2, -5) (-4)
```
this makes sense right ?

**Comment by oscarbenjamin** [comment]
Numerical evaluation of binomial from mpmath seems to be based on the other definition so SymPy's exact and numeric evaluation of binomials is currently inconsistent:
```python
In [8]: b = binomial(-1, -1, evaluate=False)

In [9]: b
Out[9]: 
⎛-1⎞
⎜  ⎟
⎝-1⎠

In [10]: b.doit()
Out[10]: 0

In [11]: b.evalf()
Out[11]: 1.00000000000000
```

**Comment by MyTrueWallet** [comment]
Good
